File: py/postcodesio_sdk.py
# Postcodesio SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import PostcodesioUtility
from core.spec import PostcodesioSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import PostcodesioBaseFeature
from features import _make_feature


class PostcodesioSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = PostcodesioUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return PostcodesioUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = PostcodesioSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def nearest(self):
        """Idiomatic facade: client.nearest.list() / client.nearest.load({"id": ...})."""
        from entity.nearest_entity import NearestEntity
        cached = getattr(self, "_nearest", None)
        if cached is None:
            cached = NearestEntity(self, None)
            self._nearest = cached
        return cached

    def Nearest(self, data=None):
        # Deprecated: use client.nearest instead.
        from entity.nearest_entity import NearestEntity
        return NearestEntity(self, data)


    @property
    def outcode(self):
        """Idiomatic facade: client.outcode.list() / client.outcode.load({"id": ...})."""
        from entity.outcode_entity import OutcodeEntity
        cached = getattr(self, "_outcode", None)
        if cached is None:
            cached = OutcodeEntity(self, None)
            self._outcode = cached
        return cached

    def Outcode(self, data=None):
        # Deprecated: use client.outcode instead.
        from entity.outcode_entity import OutcodeEntity
        return OutcodeEntity(self, data)


    @property
    def place(self):
        """Idiomatic facade: client.place.list() / client.place.load({"id": ...})."""
        from entity.place_entity import PlaceEntity
        cached = getattr(self, "_place", None)
        if cached is None:
            cached = PlaceEntity(self, None)
            self._place = cached
        return cached

    def Place(self, data=None):
        # Deprecated: use client.place instead.
        from entity.place_entity import PlaceEntity
        return PlaceEntity(self, data)


    @property
    def postcode(self):
        """Idiomatic facade: client.postcode.list() / client.postcode.load({"id": ...})."""
        from entity.postcode_entity import PostcodeEntity
        cached = getattr(self, "_postcode", None)
        if cached is None:
            cached = PostcodeEntity(self, None)
            self._postcode = cached
        return cached

    def Postcode(self, data=None):
        # Deprecated: use client.postcode instead.
        from entity.postcode_entity import PostcodeEntity
        return PostcodeEntity(self, data)


    @property
    def scottish_postcode(self):
        """Idiomatic facade: client.scottish_postcode.list() / client.scottish_postcode.load({"id": ...})."""
        from entity.scottish_postcode_entity import ScottishPostcodeEntity
        cached = getattr(self, "_scottish_postcode", None)
        if cached is None:
            cached = ScottishPostcodeEntity(self, None)
            self._scottish_postcode = cached
        return cached

    def ScottishPostcode(self, data=None):
        # Deprecated: use client.scottish_postcode instead.
        from entity.scottish_postcode_entity import ScottishPostcodeEntity
        return ScottishPostcodeEntity(self, data)


    @property
    def terminated_postcode(self):
        """Idiomatic facade: client.terminated_postcode.list() / client.terminated_postcode.load({"id": ...})."""
        from entity.terminated_postcode_entity import TerminatedPostcodeEntity
        cached = getattr(self, "_terminated_postcode", None)
        if cached is None:
            cached = TerminatedPostcodeEntity(self, None)
            self._terminated_postcode = cached
        return cached

    def TerminatedPostcode(self, data=None):
        # Deprecated: use client.terminated_postcode instead.
        from entity.terminated_postcode_entity import TerminatedPostcodeEntity
        return TerminatedPostcodeEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: lua/postcodesio_sdk.lua
-- Postcodesio SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local PostcodesioSDK = {}
PostcodesioSDK.__index = PostcodesioSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

PostcodesioSDK._make_feature = _make_feature


function PostcodesioSDK.new(options)
  local self = setmetatable({}, PostcodesioSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function PostcodesioSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function PostcodesioSDK:get_utility()
  return Utility.copy(self._utility)
end


function PostcodesioSDK:get_root_ctx()
  return self._rootctx
end


function PostcodesioSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function PostcodesioSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:nearest():list() / client:nearest():load({ id = ... })
function PostcodesioSDK:nearest(data)
  local EntityMod = require("entity.nearest_entity")
  if data == nil then
    if self._nearest == nil then
      self._nearest = EntityMod.new(self, nil)
    end
    return self._nearest
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:nearest() instead.
function PostcodesioSDK:Nearest(data)
  local EntityMod = require("entity.nearest_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:outcode():list() / client:outcode():load({ id = ... })
function PostcodesioSDK:outcode(data)
  local EntityMod = require("entity.outcode_entity")
  if data == nil then
    if self._outcode == nil then
      self._outcode = EntityMod.new(self, nil)
    end
    return self._outcode
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:outcode() instead.
function PostcodesioSDK:Outcode(data)
  local EntityMod = require("entity.outcode_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:place():list() / client:place():load({ id = ... })
function PostcodesioSDK:place(data)
  local EntityMod = require("entity.place_entity")
  if data == nil then
    if self._place == nil then
      self._place = EntityMod.new(self, nil)
    end
    return self._place
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:place() instead.
function PostcodesioSDK:Place(data)
  local EntityMod = require("entity.place_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:postcode():list() / client:postcode():load({ id = ... })
function PostcodesioSDK:postcode(data)
  local EntityMod = require("entity.postcode_entity")
  if data == nil then
    if self._postcode == nil then
      self._postcode = EntityMod.new(self, nil)
    end
    return self._postcode
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:postcode() instead.
function PostcodesioSDK:Postcode(data)
  local EntityMod = require("entity.postcode_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:scottish_postcode():list() / client:scottish_postcode():load({ id = ... })
function PostcodesioSDK:scottish_postcode(data)
  local EntityMod = require("entity.scottish_postcode_entity")
  if data == nil then
    if self._scottish_postcode == nil then
      self._scottish_postcode = EntityMod.new(self, nil)
    end
    return self._scottish_postcode
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:scottish_postcode() instead.
function PostcodesioSDK:ScottishPostcode(data)
  local EntityMod = require("entity.scottish_postcode_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:terminated_postcode():list() / client:terminated_postcode():load({ id = ... })
function PostcodesioSDK:terminated_postcode(data)
  local EntityMod = require("entity.terminated_postcode_entity")
  if data == nil then
    if self._terminated_postcode == nil then
      self._terminated_postcode = EntityMod.new(self, nil)
    end
    return self._terminated_postcode
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:terminated_postcode() instead.
function PostcodesioSDK:TerminatedPostcode(data)
  local EntityMod = require("entity.terminated_postcode_entity")
  return EntityMod.new(self, data)
end




function PostcodesioSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = PostcodesioSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return PostcodesioSDK
